database/src/service/read.rs

```diff
@@ -211,6 +226,71 @@ fn map_request(
+/// is only being used by 1 thread only to scale reads
+///
+/// <https://github.com/Cuprate/cuprate/pull/113#discussion_r1576762346>
+#[inline]
+fn thread_local<T: Send>(env: &impl Env) -> ThreadLocal<T> {
+    ThreadLocal::with_capacity(env.config().reader_threads.as_threads().get())
+}
+
+/// Depending on the backend, the table type will need `UnsafeSendable`.
+///
+/// This macro branches on the backend and uses the appropriate type.
+/// See the below `SAFETY` comment for more info.
+///
+/// # Notes
+/// This macro will early return the scope it is called from
+/// with an `Err(RuntimeError)` if there is a database error.
+macro_rules! get_tables {
+    ($env_inner:ident, $tx_ro:ident, $tables:ident) => {{
+        cfg_if::cfg_if! {
+            if #[cfg(all(feature = "redb", not(feature = "heed")))] {
+                $tables.get_or_try(|| $env_inner.open_tables($tx_ro))?
+            } else {
+                $tables.get_or_try(|| {
+                    #[allow(clippy::significant_drop_in_scrutinee)]
+                    match $env_inner.open_tables($tx_ro) {
+                        // SAFETY:
+                        // `heed`'s transactions are `Send` but `HeedTableRo` contains a `&`
+                        // to the transaction, as such, if `Send` were implemented on `HeedTableRo`
+                        // then 1 transaction could be used to open multiple tables, then sent to
+                        // other threads - this would be a soundness hole against `Sync`.
+                        //
+                        // `&T` is only `Send` if `T: Sync`.
+                        //
+                        // `heed::RoTxn: !Sync`, therefore our table
+                        // holding `&heed::RoTxn` must NOT be `Send`.
+                        //
+                        // <https://doc.rust-lang.org/std/marker/trait.Sync.html>
+                        // <https://doc.rust-lang.org/nomicon/send-and-sync.html>
+                        //
+                        // Instead, we are unsafely using `UnsafeSendable` in `service`'s reader
+                        // thread-pool as we are pairing our usage with `ThreadLocal` - only 1 thread
+                        // will ever access a transaction at a time. This is an INVARIANT.
+                        //
+                        // A `Mutex` was considered but:
+                        // - It is less performant
+                        // - It isn't technically needed for safety in our use-case
+                        // - It causes `DatabaseIter` function return issues as there is a MutexGuard object
+                        //
+                        // <https://github.com/Cuprate/cuprate/pull/113#discussion_r1581684698>
+                        Ok(tables) => Ok(unsafe { crate::unsafe_sendable::UnsafeSendable::new(tables) }),
+                        Err(e) => Err(e),
+                    }
+                })?.as_ref()
+            }
+        }
+    }};
+}
```

Comment:
How `tx_ro` + `tables` will work in the `service` reader thread-pool.

- Both get wrapped in `ThreadLocal`
- `tx_ro` is `Send`
- For `redb`, `tables` is `Send`, normal `ThreadLocal::get_or_try` is used
- For `heed`, `tables` is `!Send`, but gets wrapped in `UnsafeSendable`

usage is...

Reply:
Couldn't we just store both redb & heed tables in `UnsafeSendable` the overhead should be nothing and it means we don't need to cfg if this high up in the abstraction. Also means this can be a function I think

Reply:
Sure, although making it a function means we have to name everything:
```rust
fn get_tables<'a, EnvInner, TxRo, TxRw, Tables>(
    env_inner: &'a EnvInner,
    tx_ro: &'a TxRo,
    tables: ThreadLocal<UnsafeSendable<Tables>>,
) -> Result<&'a Tables, RuntimeError>
where
    EnvInner: crate::EnvInner<'a, TxRo, TxRw>,
    TxRo: crate::TxRo<'a>,
    TxRw: crate::TxRw<'a>,
    Tables: crate::tables::TablesIter,
{/*...*/}
```
not sure how to tell the compiler the returned value's type == `Tables` here.
```
error[E0308]: mismatched types
   --> database/src/service/read.rs:285:82
    |
246 | fn get_tables<'a, EnvInner, TxRo, TxRw, Tables>(
    |                                         ------ expected this type parameter
...
285 |                 Ok(t) => Ok(unsafe { crate::unsafe_sendable::UnsafeSendable::new(t) }),
    |                                      ------------------------------------------- ^ expected type parameter `Tables`, found associated type
    |                                      |
    |                                      arguments to this function are incorrect
    |
    = note: expected type parameter `Tables`
              found associated type `impl tables::TablesIter`
```
Can it just be a macro?

Reply:
> Can it just be a macro?

yeah, alright